llvm/include/llvm/CodeGen/VirtRegMap.h

```diff
@@ -51,6 +54,14 @@ class TargetInstrInfo;
     /// stack).
     IndexedMap<Register, VirtReg2IndexFunctor> Virt2PhysMap;
 
+    /// This maps virtual registers to their required physical register
+    /// assignments. If a VReg is present, it can only be assigned to
+    /// Virt2RequiredPhysMap[VReg] after RA finishes. This can be useful for
+    /// maintaining super-reg to sub-reg relationships. E.g. making sure that
+    /// different VRegs will still belong to a same physical super register
+    /// after allocation.
+    IndexedMap<Register, VirtReg2IndexFunctor> Virt2RequiredPhysMap;
```

Comment:
The main point is to enforce a specific assignment. Essentially this is stronger than a hint.

The reason this is here is because we run greedy multiple times, and break up large "tuple" registers into their elements in between those runs. Those sub-regs still need to be part of the same super-register, so we enforce the assignments using `setRequiredPhys`. The new VRegs, even with their required assignments, are a lot easier to handle by greedy, and can be split/spilled independently.